### PR TITLE
TYPING: add type hints to pandas\io\formats\printing.py

### PR DESCRIPTION
**Pull request by simonjayhawkins**

### CODE REVIEW

File: pandas/io/formats/printing.py [hunk 1]
```diff
@@ -3,13 +3,14 @@
...
 import sys
+from typing import Callable, Dict, Iterable, List, Optional, Sequence, Tuple, Union
 
 from pandas._config import get_option
 
 from pandas.core.dtypes.inference import is_sequence
 
 
-def adjoin(space, *lists, **kwargs):
+def adjoin(space: int, *lists: str, **kwargs) -> str:
```
Comment (jbrockmendel) [inline pandas/io/formats/printing.py]: Does this mean all the lists are strings?

File: pandas/io/formats/printing.py [hunk 2]
```diff
@@ -56,12 +57,8 @@ def justify(texts, max_len, mode="right"):
...
 
 
-def _join_unicode(lines, sep=""):
-    try:
-        return sep.join(lines)
-    except UnicodeDecodeError:
-        sep = str(sep)
-        return sep.join([x.decode("utf-8") if isinstance(x, str) else x for x in lines])
+def _join_unicode(lines: Iterable[str], sep: str = "") -> str:
+    return sep.join(lines)
```
Comment (jbrockmendel) [inline pandas/io/formats/printing.py]: Likely no longer needed?

Comment (simonjayhawkins) [reply pandas/io/formats/printing.py]: only used in two places. inline and remove?

File: pandas/io/formats/printing.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/io/formats/printing.py]: nice catch. should be List[str]. will update.

**Comment by jreback** [comment]
lgtm. @WillAyd

File: pandas/io/formats/printing.py [hunk 3]
```diff
@@ -501,16 +508,20 @@ def format_object_attrs(obj, include_dtype=True):
...
     -------
-    list
+    list of 2-tuple
 
     """
-    attrs = []
+    attrs = []  # type: List[Tuple[str, Union[str, int]]]
     if hasattr(obj, "dtype") and include_dtype:
-        attrs.append(("dtype", "'{}'".format(obj.dtype)))
+        # error: "Sequence[Any]" has no attribute "dtype"
```
Comment (WillAyd) [inline pandas/io/formats/printing.py]: Hmm this type of introspection is going to be interesting with annotations. Something to think about moving forward

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks @simonjayhawkins !

File: pandas/io/formats/printing.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/io/formats/printing.py]: see https://github.com/python/mypy/issues/1424

**Pull request merged**